src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/HttpHandlerDiagnosticListener.cs

```diff
@@ -0,0 +1,690 @@
+        /// and replace its m_WriteList arraylist field.
+        /// </summary>
+        private sealed class ConnectionArrayList : ArrayListWrapper
+        {
+            public ConnectionArrayList(ArrayList list) : base(list)
+            {
+            }
+
+            public override int Add(object value)
+            {
+                if (s_connectionType.IsInstanceOfType(value))
+                {
+                    // Replace the HttpWebRequest arraylist inside this Connection object,
+                    // which allows us to intercept each new HttpWebRequest object added under
+                    // this Connection.
+                    ArrayList originalArrayList = s_writeListField.GetValue(value) as ArrayList;
+                    HttpWebRequestArrayList newArrayList = new HttpWebRequestArrayList(originalArrayList ?? new ArrayList());
+
+                    s_writeListField.SetValue(value, newArrayList);
+                }
+
+                return base.Add(value);
+            }
+        }
+
+        /// <summary>
+        /// Helper class used for Connection.m_WriteList. The goal here is to
+        /// intercept all new HttpWebRequest objects being added to Connection.m_WriteList
+        /// and notify the listener about the HttpWebRequest that's about to send a request.
+        /// It also intercepts all HttpWebRequest objects that are about to get removed from
+        /// Connection.m_WriteList as they have completed the request.
+        /// </summary>
+        private sealed class HttpWebRequestArrayList : ArrayListWrapper
+        {
+            public HttpWebRequestArrayList(ArrayList list) : base(list)
+            {
+            }
+
+            public override int Add(object value)
+            {
+                HttpWebRequest request = value as HttpWebRequest;
+                if (request != null)
+                {
+                    s_instance.RaiseRequestEvent(request);
+                }
+
+                return base.Add(value);
+            }
+
+            public override void RemoveAt(int index)
+            {
+                HttpWebRequest request = base[index] as HttpWebRequest;
+                if (request != null)
+                {
+                    HttpWebResponse response = s_httpResponseAccessor(request);
+                    if (response != null)
+                    {
+                        s_instance.RaiseResponseEvent(request, response);
+                    }
+                }
+
+                base.RemoveAt(index);
+            }
+        }
+
+        #endregion
+
+        #region private methods
+
+        /// <summary>
+        /// Private constructor. This class implements a singleton pattern and only this class is allowed to create an instance.
+        /// </summary>
+        private HttpHandlerDiagnosticListener() : base(DiagnosticListenerName)
+        {
+        }
+
+        private void RaiseRequestEvent(HttpWebRequest request)
```

Comment:
Given this all is 'hack' code that in theory goes away, and is 'agile (it is relatively easy to update this and get it deployed), I guess I could live with this.   Still I think we overweighting the 'App Insights' scenario here.   Pretty much ANYONE using this diagnostic source (e.g. Glimpse etc) who cares about HTTP is likely to want the ID passing logic, so that argues that we should do it once for everyone here.  Also if we update this logic in the HttpClient (e.g. add interoperability with OpenTracing features (exactly what header name is used ...)), we could simply add that logic here as well and have a complete feature that works on .NET Core AND .NET Desktop.  If we delegate to the client of this hook, any updates to HttpClient need coordination with at least the AI team and probably more.   Feels like a mess.  

For things like ID and Baggage passing, de-duping is easy (if it is there already do nothing), so it seems like we should at least be doing that.

Comments?